### PR TITLE
Add listMaterializedViews and getMaterializedViews SPI

### PR DESCRIPTION
**Pull request by raunaqmorarka**
Used listMaterializedViews to list MV names for information_schema.tables, system.jdbc.tables and SHOW TABLES
Used getMaterializedViews to list MV columns in information_schema.columns, system.jdbc.columns and SHOW COLUMNS

Fixed DESCRIBE, SHOW COLUMNS, information_schema.columns, system.jdbc.columns for MVs in Iceberg.

Table type used for MVs in information_schema.tables is left unchanged as BASE TABLE.
MVs are not listed in information_schema.views.
getMaterializedViews can be used in future to introduce information_schema.trino_materialized_views or system.metadata.materialized_views for showing detailed information about materialized view definitions.

### CODE REVIEW

**Comment by findepi** [comment]
@raunaqmorarka i am adding more tests, inclduding a bit of information_schema coverage for MVs in https://github.com/trinodb/trino/pull/8117
Can you please check how your PR interacts with the tests added there?

**Comment by raunaqmorarka** [comment]
> @raunaqmorarka i am adding more tests, inclduding a bit of information_schema coverage for MVs in #8117
> Can you please check how your PR interacts with the tests added there?

Looked at it, since i've tested similar cases in this PR, I think it should result in the `TODO` cases of that PR getting solved.

**Comment by raunaqmorarka** [comment]
@findepi I've rebased this to master and added changes to BaseConnectorTest#testMaterializedView

File: core/trino-main/src/main/java/io/trino/sql/rewrite/ShowQueriesRewrite.java [hunk 1]
```diff
@@ -389,13 +389,20 @@ protected Node visitShowColumns(ShowColumns showColumns, Void context)
...
-            Optional<ConnectorViewDefinition> view = metadata.getView(session, tableName);
-            Optional<TableHandle> tableHandle = metadata.getTableHandle(session, tableName);
-            if (view.isEmpty() && tableHandle.isEmpty()) {
+
+            boolean isMaterializedView = metadata.getMaterializedView(session, tableName).isPresent();
+            boolean isView = false;
+            Optional<TableHandle> tableHandle = Optional.empty();
+            if (!isMaterializedView) {
+                isView = metadata.getView(session, tableName).isPresent();
+                tableHandle = metadata.getTableHandle(session, tableName);
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/sql/rewrite/ShowQueriesRewrite.java:398]: i see you're saving `metadata.getView` when `isMaterializedView`
why aren't you saving `metadata.getTableHandle` call when something `isView`?

File: core/trino-main/src/main/java/io/trino/security/AccessControl.java [hunk 2]
```diff
@@ -206,7 +206,7 @@
     void checkCanShowTables(SecurityContext context, CatalogSchemaName schema);
 
     /**
-     * Filter the list of tables and views to those visible to the identity.
+     * Filter the list of tables, materialized views and views to those visible to the identity.
```
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/security/AccessControl.java:209]: I'm using filterTables in listMaterializedViews similar to listViews in MetadataListing

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMaterializedViews.java [hunk 3]
```diff
@@ -103,6 +103,22 @@ public void testSystemJdbc()
...
                 "VALUES ('iceberg', 'tpch', 'system_jdbc_test', 'TABLE')");
 
+        assertThat(query(
+                "SELECT table_cat, table_schem, table_name, column_name, type_name " +
+                        "FROM system.jdbc.columns " +
+                        "WHERE table_schem LIKE '%tpch%'"))
+                .skippingTypesCheck()
+                .containsAll(
+                        "VALUES ('iceberg', 'tpch', 'system_jdbc_test', '_bigint', 'bigint'), " +
+                                "('iceberg', 'tpch', 'system_jdbc_test', '_date', 'date')");
```
Comment (findepi) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMaterializedViews.java:113]: let's add this in `BCT.testMaterializedView`, maybe without type names, for the sake of connectors which e.g. do not suppport bigint

File: testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java [hunk 4]
```diff
@@ -437,13 +438,26 @@ public void testMaterializedView()
                 "VALUES '" + viewName + "'");
 
         // column listing
-        checkShowColumnsForMaterializedView(viewName);
+        MaterializedResult expectedColumns = resultBuilder(getSession(), VARCHAR, VARCHAR, VARCHAR, VARCHAR)
+                .row("nationkey", "bigint", "", "")
+                .row("name", "varchar", "", "")
+                .row("regionkey", "bigint", "", "")
+                .row("comment", "varchar", "", "")
```
Comment (findepi) [inline testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java:445]: Connector can return comments or something, and we're not interested in these.
you can use `QueryAssert#projected` to ignore them in assertions.

Also, the column types can vary between connectors, when eg they don't support `bigint` (like oracle)

File: testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java [hunk 5]
```diff
@@ -460,35 +474,6 @@ public void testMaterializedView()
         assertUpdate("DROP MATERIALIZED VIEW " + viewName);
     }
 
-    // TODO inline when all implementations fixed
-    protected void checkShowColumnsForMaterializedView(String viewName)
```
Comment (findepi) [inline testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java:470]: it could be good to have a commit that does the inline, separate from other changes

File: core/trino-main/src/main/java/io/trino/connector/system/jdbc/ColumnJdbcTable.java [hunk 6]
```diff
@@ -267,7 +273,9 @@ public RecordCursor cursor(ConnectorTransactionHandle transactionHandle, Connect
                     QualifiedTablePrefix tablePrefix = tableFilter.isPresent()
                             ? new QualifiedTablePrefix(catalog, schema, tableFilter.get())
                             : new QualifiedTablePrefix(catalog, schema);
-                    Set<SchemaTableName> tables = listTables(session, metadata, accessControl, tablePrefix);
+                    Set<SchemaTableName> tables = union(
+                            listTables(session, metadata, accessControl, tablePrefix),
+                            listMaterializedViews(session, metadata, accessControl, tablePrefix));
```
[the remark this answers is not in the document]
Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/connector/system/jdbc/ColumnJdbcTable.java:278]: This PR doesn't change `listTables` implementation yet. This will be subsequent PR

[the remark this answers is not in the document]
Comment (findepi) [reply core/trino-main/src/main/java/io/trino/connector/system/jdbc/ColumnJdbcTable.java:278]: Then we at least need to reach agreement for the general direction. see https://github.com/trinodb/trino/pull/8131#issuecomment-851179231

File: core/trino-main/src/main/java/io/trino/connector/informationschema/InformationSchemaMetadata.java [hunk 7]
```diff
@@ -353,7 +353,10 @@ public static boolean isTablesEnumeratingTable(InformationSchemaTable table)
                     .flatMap(prefix -> tables.get().stream()
                             .filter(this::isLowerCase)
                             .map(table -> new QualifiedObjectName(catalogName, prefix.getSchemaName().get(), table)))
-                    .filter(objectName -> !isColumnsEnumeratingTable(informationSchemaTable) || metadata.getTableHandle(session, objectName).isPresent() || metadata.getView(session, objectName).isPresent())
+                    .filter(objectName -> !isColumnsEnumeratingTable(informationSchemaTable) ||
+                            metadata.getMaterializedView(session, objectName).isPresent() ||
+                            metadata.getTableHandle(session, objectName).isPresent() ||
+                            metadata.getView(session, objectName).isPresent())
```
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/connector/informationschema/InformationSchemaMetadata.java:359]: Iceberg getTableHandle throws an exception if it is called for an MV
```
if (!isIcebergTable(hiveTable.get())) {
            throw new UnknownTableTypeException(tableName);
}
```
Calling `getMaterializedView.isPresent` first side-steps this exception.
I'm not entirely sure if iceberg behaviour is the right one though.
Let me know if we should change iceberg getTableHandle to return empty instead or put a try/catch around getTableHandle call here.

File: core/trino-main/src/main/java/io/trino/metadata/MetadataListing.java [hunk 8]
```diff
@@ -117,6 +117,14 @@ private MetadataListing() {}
                 .collect(toImmutableMap(Entry::getKey, Entry::getValue));
     }
 
+    public static Set<SchemaTableName> listMaterializedViews(Session session, Metadata metadata, AccessControl accessControl, QualifiedTablePrefix prefix)
+    {
+        Set<SchemaTableName> tableNames = metadata.listMaterializedViews(session, prefix).stream()
+                .map(QualifiedObjectName::asSchemaTableName)
+                .collect(toImmutableSet());
+        return accessControl.filterTables(session.toSecurityContext(), prefix.getCatalogName(), tableNames);
```
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/metadata/MetadataListing.java:125]: I think it makes sense to rename `AccessControl#filterTables` to `filterRelations` at this point.

File: core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java [hunk 9]
```diff
@@ -600,6 +600,9 @@ public ColumnMetadata getColumnMetadata(Session session, TableHandle tableHandle
 
         Optional<QualifiedObjectName> objectName = prefix.asQualifiedObjectName();
         if (objectName.isPresent()) {
+            if (getMaterializedView(session, objectName.get()).isPresent()) {
+                return ImmutableList.of();
```
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java:604]: Iceberg getTableHandle throws an exception if it is called for an MV
```
if (!isIcebergTable(hiveTable.get())) {
            throw new UnknownTableTypeException(tableName);
}
```
isExistingRelation seems to be accepting views based on the understanding that views should be part of listTables, another case of different interpretation of API :)

File: core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java [hunk 10]
```diff
@@ -671,6 +671,20 @@ private boolean isExistingRelation(Session session, QualifiedObjectName name)
                     }
                     tableColumns.put(entry.getKey(), columns.build());
                 }
+
+                // if view and materialized view names overlap, the materialized view wins
```
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java:675]: I think winning here just means overwriting any existing value in tableColumns map.
Since the priority here is MV > View > Table, we're consistent with StatementAnalyzer.Visitor#visitTable ?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 11]
```diff
@@ -1028,6 +1020,17 @@ private boolean isMaterializedView(Table table)
                 && table.getParameters().containsKey(STORAGE_TABLE);
     }
 
+    @Override
+    public List<SchemaTableName> listMaterializedViews(ConnectorSession session, Optional<String> schemaName)
+    {
+        // Iceberg does not support VIEWs yet, therefore all views in HMS are materialized views
+        return schemaName.map(Collections::singletonList)
+                .orElseGet(metastore::getAllDatabases).stream()
+                .flatMap(schema -> metastore.getAllViews(schema).stream()
```
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:1029]: Agreed, this seems to be an existing bug, created https://github.com/trinodb/trino/pull/8153 for it

File: core/trino-main/src/main/java/io/trino/connector/informationschema/InformationSchemaMetadata.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (findepi) [reply core/trino-main/src/main/java/io/trino/connector/informationschema/InformationSchemaMetadata.java:359]: Thanks for taking care of that

File: core/trino-main/src/main/java/io/trino/metadata/MetadataListing.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/metadata/MetadataListing.java:125]: The rename incurs a bunch of changes in APIs, I'll do that as a follow up if necessary

File: core/trino-main/src/main/java/io/trino/connector/informationschema/InformationSchemaMetadata.java [hunk 12]
```diff
@@ -353,7 +353,10 @@ public static boolean isTablesEnumeratingTable(InformationSchemaTable table)
                     .flatMap(prefix -> tables.get().stream()
                             .filter(this::isLowerCase)
                             .map(table -> new QualifiedObjectName(catalogName, prefix.getSchemaName().get(), table)))
-                    .filter(objectName -> !isColumnsEnumeratingTable(informationSchemaTable) || metadata.getTableHandle(session, objectName).isPresent() || metadata.getView(session, objectName).isPresent())
+                    .filter(objectName -> !isColumnsEnumeratingTable(informationSchemaTable) ||
```
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/connector/informationschema/InformationSchemaMetadata.java:357]: This is covered by BCT column listing tests

File: core/trino-main/src/main/java/io/trino/connector/system/jdbc/ColumnJdbcTable.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/connector/system/jdbc/ColumnJdbcTable.java]: I've updated this PR to avoid explicitly calling listMVs in information_schema or system.jdbc
So the code in ColumnJdbc, InformationSchemaMetadata and TableJdbc remains unchanged

File: core/trino-main/src/main/java/io/trino/connector/system/jdbc/TableJdbcTable.java [hunk 13]
```diff
@@ -103,7 +105,9 @@ public RecordCursor cursor(ConnectorTransactionHandle transactionHandle, Connect
             QualifiedTablePrefix prefix = tablePrefix(catalog, schemaFilter, tableFilter);
 
             Set<SchemaTableName> views = listViews(session, metadata, accessControl, prefix);
-            for (SchemaTableName name : listTables(session, metadata, accessControl, prefix)) {
+            Set<SchemaTableName> tables = listTables(session, metadata, accessControl, prefix);
+            Set<SchemaTableName> materializedViews = listMaterializedViews(session, metadata, accessControl, prefix);
+            for (SchemaTableName name : union(tables, materializedViews)) {
```
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/connector/system/jdbc/TableJdbcTable.java]: Created https://github.com/trinodb/trino/issues/8167 for BCT coverage of system.jdbc
With current changes, I'm able to leave this class unchanged

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 14]
```diff
@@ -1032,6 +1028,17 @@ private boolean isMaterializedView(Table table)
                 && table.getParameters().containsKey(STORAGE_TABLE);
     }
 
+    @Override
+    public List<SchemaTableName> listMaterializedViews(ConnectorSession session, Optional<String> schemaName)
```
Comment (sopel39) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:1032]: Let's add test in TestIcebergMetadataListing.java..

Should https://github.com/trinodb/trino/pull/8153 land first?

File: testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java [hunk 15]
```diff
@@ -481,14 +506,15 @@ protected void checkInformationSchemaTablesForPointedQueryForMaterializedView(St
     }
 
     // TODO inline when all implementations fixed
-    protected void checkShowColumnsForMaterializedView(String schemaName, String viewName)
+    protected void checkShowColumnsForMaterializedView(String viewName)
     {
         assertThat(query("SHOW COLUMNS FROM " + viewName))
+                .projected(0)
```
Comment (sopel39) [inline testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java]: would matching column type be hard to do instead?

**Review by sopel39 — APPROVED** [review]
% small comments.

Should https://github.com/trinodb/trino/pull/8153 land first?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java (hunk 14 above)
Comment (raunaqmorarka) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:1032]: yes, let's land that one first

File: testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java (hunk 15 above)
Comment (raunaqmorarka) [reply testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java]: Since column types can vary between connectors, trying to do that in BCT could mean overriding implementations in individual connectors

**Comment by sopel39** [comment]
please rebase on top of https://github.com/trinodb/trino/pull/8153

**Comment by raunaqmorarka** [comment]
> please rebase on top of #8153

done

File: core/trino-main/src/main/java/io/trino/connector/informationschema/InformationSchemaPageSource.java [hunk 16]
```diff
@@ -280,6 +280,7 @@ private void addTablesRecords(QualifiedTablePrefix prefix)
     {
         Set<SchemaTableName> tables = listTables(session, metadata, accessControl, prefix);
         Set<SchemaTableName> views = listViews(session, metadata, accessControl, prefix);
+        // TODO define a type for materialized views
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/connector/informationschema/InformationSchemaPageSource.java:283]: could you put link to ticket for TODO?

File: testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java (hunk 15 above)
Comment (sopel39) [reply testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java]: Please add a comment about it

**Review by sopel39 — APPROVED** [review]
lgtm % minor comments

File: core/trino-main/src/main/java/io/trino/connector/informationschema/InformationSchemaPageSource.java (hunk 16 above)
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/connector/informationschema/InformationSchemaPageSource.java:283]: Created https://github.com/trinodb/trino/issues/8207

File: core/trino-main/src/main/java/io/trino/connector/informationschema/InformationSchemaPageSource.java [hunk 17]
```diff
@@ -280,6 +280,7 @@ private void addTablesRecords(QualifiedTablePrefix prefix)
     {
         Set<SchemaTableName> tables = listTables(session, metadata, accessControl, prefix);
         Set<SchemaTableName> views = listViews(session, metadata, accessControl, prefix);
+        // TODO (https://github.com/trinodb/trino/issues/8207) define a type for materialized views
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/connector/informationschema/InformationSchemaPageSource.java:283]: nit: this should be separate commit
